### PR TITLE
Genesis web3

### PR DESCRIPTION
**Pull request by yglukhov**

### CODE REVIEW

File: beacon_chain/genesis.nim [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  let web3 = newWeb3(provider)
+
+  var contractAddress = Address.fromHex(conf.depositContractAddress)
+  var defaultAccount: Address
+  var ns = web3.contractSender(DepositContract, contractAddress, defaultAccount)
+
+  var deposits = DepositCollector()
+  deposits.queue = newAsyncQueue[QueueElement]()
+
+  let s = await ns.subscribe(DepositEvent, %*{"fromBlock": "0x0"}) do(pubkey: Bytes48, withdrawalCredentials: Bytes32, amount: Bytes8, signature: Bytes96, merkleTreeIndex: Bytes8, j: JsonNode):
```
Comment (zah) [inline beacon_chain/genesis.nim]: How would this work when you connect to a network where the genesis has already happened? The `fromBlock` takes care of this?

Also, we should keep monitoring the contract past the genesis event for new deposits

Comment (yglukhov) [reply beacon_chain/genesis.nim]: > The `fromBlock` takes care of this?

yes

> Also, we should keep monitoring the contract past the genesis event for new deposits

True, I still haven't looked into that yet.

**Comment by yglukhov** [comment]
> Hide the new functionality behind a switch (we can then merge it quite quickly)

Please see the "Make web3 way of getting genesis optional" commit. TLDR omitting `--depositWeb3Url` will trigger the old way of doing things.

**Comment by yglukhov** [comment]
https://github.com/status-im/nim-web3/pull/2 should be merged first though

**Comment by arnetheduck** [comment]
It's feature that we can launch without eth1 - something that we should keep almost as a primary or at least equal option - ie that we somehow abstract genesis / initial validator set construction so that we can support multiple methods.

Most immediately this will be needed for the various cross-client testing tools, but also when launching custom testnets that people will use when running things locally / private nets etc.

**Comment by zah** [comment]
Why is this still not being merged? We need to resolve the CI issues by creating a web3 nimble package?

**Comment by zah** [comment]
@arnetheduck, in the latest version, @yglukhov has implemented a command-line switch enabling the new behavior.

**Comment by yglukhov** [comment]
> We need to resolve the CI issues by creating a web3 nimble package?

@zah yes, please see https://github.com/nim-lang/packages/pull/1133

**Comment by zah** [comment]
I've merged the Nimble PR

**Pull request merged**